### PR TITLE
Zrobot and zdb restart

### PR DESCRIPTION
**Pull request by Dinaamagdy**

### CODE REVIEW

File: test/failures_testsuite/base_test.py [hunk 1]
```diff
@@ -75,7 +75,7 @@ def tearDownClass(cls):
         """
         self = cls()
         self._delete_directory(directory='tmp')
-        self.s3.failures.tlog_up()
+        #self.s3.failures.tlog_up()
```
Comment (john-kheir) [inline test/failures_testsuite/base_test.py]: not sure why did u remove this ?

File: test/failures_testsuite/testcases.py [hunk 2]
```diff
@@ -202,3 +201,18 @@ def test008_stop_tlog_upload_download(self):
         md5_after = self.download_file(file_name=self.file_name, keep_trying=True)
         self.assertEqual(md5_after, md5_before)
 
+    def test009_zdb_down(self):
+        """
+        - down zdb process  and make sure it will restart automatically.
```
Comment (john-kheir) [inline test/failures_testsuite/testcases.py]: replace 'down' --> 'kill' .. same for test10

File: test/failures_testsuite/utils/failures.py [hunk 3]
```diff
@@ -72,6 +73,43 @@ def minio_process_down(self, timeout):
...
+                n +=1
+            except StateCheckError:
+                continue
+            logger.info('stop %s  process on node %s', zdb.name, namespace['node'])
+            zdb_node = j.clients.zos.get(zdb.name,data={"host": namespace['url'][7:-5]})
+            zdb_cont_client = zdb_node.containers.get("zerodb_{}".format(zdb.name))
+            zdb_node.client.container.terminate(zdb_cont_client.id)
+            logger.info('zdb process killed')
+            start = time.time()
+            while (start + timeout) > time.time():
```
Comment (john-kheir) [inline test/failures_testsuite/utils/failures.py]:
```start = time.time()
            while (start + timeout) > time.time():
                try:
                    zdb_node.containers.get("zerodb_{}".format(zdb.name))
                    end = time.time()
                    duration = end - start 
                    logger.info("zdb took %s sec to restart" % duration)
                    return True
                except LookupError:
                    continue
return False```

File: test/failures_testsuite/utils/failures.py [hunk 4]
```diff
@@ -72,6 +73,43 @@ def minio_process_down(self, timeout):
...
+            zdb = robot.services.get(name=ns.data['data']['zerodb'])
+            try:
+                zdb.state.check('status', 'running', 'ok')
+                n +=1
+            except StateCheckError:
+                continue
+            logger.info('stop %s  process on node %s', zdb.name, namespace['node'])
+            zdb_node = j.clients.zos.get(zdb.name,data={"host": namespace['url'][7:-5]})
+            zdb_cont_client = zdb_node.containers.get("zerodb_{}".format(zdb.name))
+            zdb_node.client.container.terminate(zdb_cont_client.id)
```
Comment (john-kheir) [inline test/failures_testsuite/utils/failures.py]: i'm not sure if they want the container down or the process .. so please check this .. Also anyways i think u need to make sure that the process is running inside the container not just that the container is up... same scenario for robot test case

File: test/failures_testsuite/utils/failures.py [hunk 5]
```diff
@@ -206,6 +244,45 @@ def kill_tlog(self):
...
+            return
+            
+        if not node_addr:
+            farm_name = s3.service.data['data']['farmerIyoOrg']
+            capacity = j.clients.threefold_directory.get(interactive=False)
+            nodes_data = capacity.api.ListCapacity(query_params={'farmer': farm_name})[1].json()
+            for _ in range(len(nodes_data)):
+                node_addr = random.choice(nodes_data)["robot_address"][7:-5]
+                node = j.clients.zos.get("zrobot", data={"host":node_addr})
+                try:
```
Comment (john-kheir) [inline test/failures_testsuite/utils/failures.py:265]: this "try, except" should be outside of the "if" section

File: test/failures_testsuite/base_test.py (hunk 1 above)
Comment (Dinaamagdy) [reply test/failures_testsuite/base_test.py]: it didn't work and raise error , islam need to edit it

File: test/failures_testsuite/testcases.py (hunk 2 above)
Comment (Dinaamagdy) [reply test/failures_testsuite/testcases.py]: done

File: test/failures_testsuite/base_test.py (hunk 1 above)
Comment (0xIslamTaha) [reply test/failures_testsuite/base_test.py]: I removed it

File: test/failures_testsuite/testcases.py [hunk 6]
```diff
@@ -206,3 +205,34 @@ def test008_stop_tlog_upload_download(self):
         md5_after = self.download_file(file_name=self.file_name, keep_trying=True)
         self.assertEqual(md5_after, md5_before)
 
+    def test009_zdb_kill(self):
+        """
+        - kill zdb process  and make sure it will restart automatically.
+        """
+        self.logger.info("Upload File.")
```
Comment (0xIslamTaha) [inline test/failures_testsuite/testcases.py]: remove it, upload method has its own logs

File: test/failures_testsuite/testcases.py [hunk 7]
```diff
@@ -206,3 +205,34 @@ def test008_stop_tlog_upload_download(self):
...
+        file_name = self.upload_file()
+        md5_before = file_name
```
Comment (0xIslamTaha) [inline test/failures_testsuite/testcases.py]:
```suggestion
        md5_before = self.upload_file()
```

File: test/failures_testsuite/utils/failures.py [hunk 8]
```diff
@@ -72,6 +73,47 @@ def minio_process_down(self, timeout):
...
+                continue
+            logger.info('stop %s  process on node %s', zdb.name, namespace['node'])
+            zdb_node = j.clients.zos.get(zdb.name,data={"host": namespace['url'][7:-5]})
+            zdb_cont_client = zdb_node.containers.get("zerodb_{}".format(zdb.name))
+            job_id = "zerodb.{}".format(zdb.name)
+            result = zdb_cont_client.client.job.kill(job_id, signal=signal.SIGINT)
+            if not result:
+                logger.info("zerodb job not exist")
+                return False
+            logger.info('zdb process killed')
```
Comment (0xIslamTaha) [inline test/failures_testsuite/utils/failures.py]:
```suggestion
            logger.info('zdb process has been killed')
```

File: test/failures_testsuite/testcases.py [hunk 9]
```diff
@@ -206,3 +205,34 @@ def test008_stop_tlog_upload_download(self):
...
+    def test010_zrobot_kill(self):
+        """
+        - kill zrobot process and make sure it will restart automatically.
+        """
+        self.logger.info("Upload File.")
+        file_name = self.upload_file()
+        md5_before = file_name
+
+        self.logger.info('kill zrobot process and make sure it will restart automatically')
+        flag = self.s3.failures.Kill_node_robot_process()
```
Comment (0xIslamTaha) [inline test/failures_testsuite/testcases.py]: kill one of the s3 nodes, u can kill one of namespaces or tlog nodes

File: test/failures_testsuite/testcases/test02_zdb.py [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+from base_test import BaseTest
+import unittest, time
+
+class ZDBFailures(BaseTest):    
+    def tearDown(self):
+        super().tearDown()
+
+    def test001_zdb_kill(self):
+        """
+        - kill zdb process  and make sure it will restart automatically.
```
Comment (0xIslamTaha) [inline test/failures_testsuite/testcases/test02_zdb.py:12]: upload, kill, download 
update steps

File: test/failures_testsuite/testcases/test03_zrobot.py [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+from base_test import BaseTest
+import unittest, time
+
+class ZROBOTFailures(BaseTest):      
+    def tearDown(self):
+        super().tearDown()
+
+    def test001_zrobot_kill(self):
+        """
+        - kill zrobot process and make sure it will restart automatically.
```
Comment (0xIslamTaha) [inline test/failures_testsuite/testcases/test03_zrobot.py]: update steps

**Review by 0xIslamTaha — APPROVED** [review]

**Pull request merged**